File: grant_clean_js/6.js
const cheerio = require('cheerio');
const moment = require('moment');
function getResearcherNew(cheer) {
    try {
        if(cheer("section.block").find("p").text().trim()== "")
        return [null, null];
        cheer = cheer("section.block").find("p").html();
    }
    catch(error) {
        return [null, null];
    }

    var researcher = null;
    var institution = null;
    cheer = cheerio.load(cheer.toString().replace(/<br\/>/g, ";"));

    for (row of cheer.text().split(";")) {
        if (row.trim() == "" ){
            continue;
        }
        var cand = row.split(":");
        if (cand.length == 2 ){
            if (cand[0].trim() == "RST") {
                researcher = cand[1].trim();
            }
            else if(cand[0].trim() == "Etablissement Coordinateur") {
                institution = cand[1].trim();
            }
        }
    }
    return [researcher, institution];
}
    

function getTitle(cheer) {
    try {
        if(cheer("section.content-style").find("h1").text().trim() == "")
            return null;
        return cheer("section.content-style").find("h1").text().trim();
        // return cheer.find("section", {"class" { "content-style"}).find("h1").text().trim();
    }
    catch(error) {
        return null;
    }
    return null;
}

function getInstitution(cheer) {
    var institutionCand = [];
    try {
        cheer("section.block.block-info").each(function(i, elem){
            institutionCand.push(cheer(this));
        });
         //cheer.find_all("section", {"class" { "block block-info"})
    }
    catch(error) {
        return null;
    }

    try {
        for (row of institutionCand) {
            if (row("h2").first().text().trim() == "Partner") {
                var institution = row("div.block__content").find("p").text();
                let tmpInst = [];
                for(ins of institution.split("\n")){
                    if(ins.trim() != ""){
                        tmpInst.push(ins.trim());
                    }
                }
                institution = tmpInst.join(";");
                //institution = ";".join([ins.trim() for ins of institution.split("\n") if ins.trim() != ""])
                return institution;
            }
        }
    }
    catch(error) { 
        return null;
    }
    return null;
}

function getDescription(cheer) {
    var description;
    try {
        if(description = cheer("section.content-style").first().text().trim() == "")
            return null;
        description = cheer("section.content-style").first();
    }
    catch(error) {
        return null;
    }
    description("section").each(function(i, elem){
        description(this).remove();
    });
    description("h1").each(function(i, elem){
        description(this).remove();
    });
    description("div.categories").each(function(i, elem){
        description(this).remove();
    });
    // [s.extract() for s of description.find_all("section")]
    // [s.extract() for s of description.find_all("h1")]
    // [s.extract() for s of description.find_all("div", {"class" { "categories"})]

    return description.text().trim();
}
function getAwardAmount(cheer) {
    var cands = [];
    try {
        cheer("section.block.block-info").each(function(i, elem){
            cands.push(cheer(this));
        });
        // cands = cheer.find_all("section", {"class" { "block block-info"})
    }
    catch(error) {
        return [null, null];
    }

    try {
        for (row of cands) {
            if (row("h2").first().text().trim() == "Partner") {
                var strongs = [];
                row("strong").each(function(i,elem){
                    strongs.push(row(this));
                })
                // row.find_all("strong")
                for (strong of strongs) {
                    if (strong.text().indexOf("ANR") != -1) {
                        var currency =  strong.indexOf("euros") != -1 ? "EUR" : null;
                        //"EUR" if strong.find("euros") != -1 else null
                        var awardAmount = strong.text().replace(/[^\d\.]/g,"");
                        //re.sub(r"[^\d\.]", "", strong.text())
                        try {
                            awardAmount = parseFloat(awardAmount);
                        }
                        catch(error) {
                            return [null, null];
                        }
                        return awardAmount, currency
                    }
                }
            }
        }
    }
    catch(error) { 
        return [null, null];
    }
    return [null, null];
}

function getAwardAmountNew(cheer) {
    try {
        if(cheer("section.block").first().text().trim() == "")
            return [null, null];
        cheer = cheer("section.block").first();
        //cheer.find("section", {"class" { "block"})
    }
    catch(error) {
        return [null, null];
    }

    try {
        var strongs = [];
        cheer("strong").each(function(i,elem){
            strongs.push(cheer(this));
        });
        // cheer.find_all("strong")
        for (strong of strongs) {
            if (strong.text().indexOf("ANR") != -1) {
                var currency = strong.text().indexOf("euros") != -1 ? "EUR" : null;
                //"EUR" if strong.find("euros") != -1 else null
                var awardAmount = strong.text().replace(/[^\d\.]/,"");
                //re.sub(r"[^\d\.]", "", strong.text())
                try {
                    awardAmount = parseFloat(awardAmount);
                }
                catch(error) {
                    return [null, null];
                }
                return [awardAmount, currency];
            }
        }
    }
    catch(error) { 
        return [null, null];
    }
    return [null, null];
}

function getDateNew(cheer) {
    try {
        if(cheer("section.block").first().text().trim() == "")
            return  [null, null];
        cheer = cheer("section.block").first();
    }
    catch(error) {
        return [null, null];
    }

    try {
        // strongs = cheer.find_all("strong")
        var strongs = [];
        cheer("strong").each(function(i,elem){
            strongs.push(cheer(this));
        });
        for (strong of strongs) {
            if (strong.text().indexOf("Investissement") != -1 ){
                var text = strong.text().trim();
                text = text.split("période de")[text.split("période de").length-1]; //[-1]
                var cal = {
                    "janvier" : "January",
                    "février" : "February",
                    "mars" : "March",
                    "avril" : "April",
                    "mai" : "May",
                    "juin" : "June",
                    "juillet" : "July",
                    "aout" : "August",
                    "septembre" : "September",
                    "octobre" : "October",
                    "novembre" : "November",
                    "décembre" : "December",
                };
                var tmp = [];
                for (row of text.split(" ")) {
                    if (row.toLowerCase().trim() in cal) {
                        tmp.push(cal[row.lower().trim()]);
                    }
                    else {
                        tmp.push(row);
                    }
                }
                var text = tmp.join(" ");//" ".join(tmp)
                //text = [t.trim() for t of text.split("à")]
                let txt = [];
                for(t of text.split("à")){
                    txt.push(t.trim());
                }
                if (txt.length > 1) {
                    if (txt[0] == "") {
                        return [null, null];
                    }
                    var d = txt[0].split(" ");
                    txt[0] = moment().year(d[1]).month(d[0]).format("YYYY-MM-DD");
                    //datetime.datetime.strptime(txt[0], "%B %Y").date()
                    d = txt[1].split(" ");
                    txt[1] = moment().year(d[1]).month(d[0]).format("YYYY-MM-DD");
                    //datetime.datetime.strptime(txt[1], "%B %Y").date()
                    return [(txt[0]).toString(), (txt[1]).toString()];
                }
            }
        }
    }
    catch(error) { 
        return [null, null];
    }
    return [null, null];
}

function getDate(cheer) {
    var cands = [];
    try {
        cands = cheer("section.block.block-info").each(function(i, elem){
            cands.push(cheer(this));
        });
        //cheer.find_all("section", {"class" { "block block-info"})
    }
    catch(error) {
        return [null, null];
    }

    try {
        for (row of cands ){
            if (row("h2").first().text().trim() == "Partner" ){
                // strongs = row.find_all("strong")
                var strongs = [];
                row("strong").each(function(i,elem){
                    strongs.push(row(this));
                });
                for (strong of strongs) {
                    if (strong.text().indexOf("project") != -1 ){
                        var text = strong.text().trim();
                        text = text.split(":")[text.split(":").length-1];
                        // text = [t.trim() for t of text.split("-")]
                        var txt = [];
                        for(t of text.split("-")){
                            txt.push(t.trim());
                        }
                        if (txt.length > 1) {
                            if (txt[0] == "") {
                                return [null, null];
                            }
                            var d = txt[0].split(" ");
                            txt[0] = moment().year(d[1]).month(d[0]);//.format("YYYY-MM-DD");
                            // txt[0] = datetime.datetime.strptime(text[0], "%B %Y").date()
                            var tmp = txt[0];
                            var pos = txt[1].indexOf(" Months");
                            
                            if (pos != -1 ){
                                var year_save =  parseInt(parseInt(txt[1].substr(0,pos)) / 12);//int(int(txt[1][{ pos]) / 12)
                                var month_save = parseInt(parseInt(txt[1].substr(0,pos)) % 12)//int(int(txt[1][{ pos]) % 12)
                                var year_save = parseInt(tmp.year()) + year_save + parseInt((parseInt(tmp.month()) + 1 + month_save - 1) / 12)
                                month_save = (month_save + tmp.month() + 1) % 12
                                if (month_save == 0) {
                                    month_save = 12;
                                    tmp = tmp.format("YYYY-MM-DD").toString().replace(d[1],year_save.toString()).replace(tmp.toString().substr(5,2),month_save.toString);                                    
                                        // txt[1] = tmp.replace(year=year_save, month=month_save)
                                    txt[1] = tmp;
                                }
                                return [(txt[0]).toString(), (txt[1]).toString()];
                            }
                        }
                    }
                }
            }
        }
    }
    catch(error) { 
        return [null, null];
    }
    return [null, null];
}


function getEndYear(endDate) {
    if (endDate == null || endDate.length == 0) {
        return null;
    }
    if (endDate.length > 4) {
        return endDate.substr(0,4);
        // return endDate[ : 4]
    }
    return null;
}

function getProjectId(projectUrl) {
    if (projectUrl == null) {
        return null;
    }
    if (projectUrl.indexOf("-") != -1 ){
        return projectUrl.toString().substr(projectUrl.indexOf("-")+1).trim();  //[projectUrl.find("-") + 1 :].trim()
    }
    return null;
}
    


function cleanHtml(html){
    return html.toString().replace(/\n|\t|\s+/g," ");
    //re.sub(r"\n|\t|\s+", " ", str(cheer))
}


async function parse(features){
    var html = features['html'];
    if(html == null){
        return features;
    }
    html = cleanHtml(html);
    let cheer = cheerio.load(html, {decodeEntities : false});
    if (features["project_url"].indexOf("ProjetIA") != -1) {
        features["title"]    = getTitle(cheer);
        features["title_en"] = features["title"];

        features["project_id"]          = getProjectId(features["project_url"]);
        features["project_id_original"] = features["project_id"];
        var award = getAwardAmountNew(cheer);
        features["award_amount"] = award[0];
        features["currency"] = award[1];
        var res = getResearcherNew(cheer);
        features["researcher"] =res[0];
        features["institution"] = res[1];
        var dates = getDateNew(cheer);
        features["start_date"] = dates[0];
        features["end_date"] = dates[1];

        features["description"]  = getDescription(cheer)
        features["description_raw"] = features["description"];
        features["description_en"] = features["description"];

        features["start_year"] = getEndYear(features["start_date"]);
        features["end_year"] = getEndYear(features["end_date"]);
    }
    else {
        features["title"]    = getTitle(cheer);
        features["title_en"] = features["title"];

        features["project_id"]          = getProjectId(features["project_url"]);
        features["project_id_original"] = features["project_id"];
        
        features["institution"] = getInstitution(cheer);
        features["institution_en"] = features["institution"];
        var awards = getAwardAmount(cheer);
        features["award_amount"] = awards[0];
        features["currency"]  = awards[1];
        var date = getDate(cheer);
        features["start_date"] = date[0];
        features["end_date"] = date[1];

        features["description"]  = getDescription(cheer);
        features["description_raw"] = features["description"];
        features["description_en"] = features["description"];

        features["start_year"] = getEndYear(features["start_date"]);
        features["end_year"] = getEndYear(features["end_date"]);
    }
    return features;
}

module.exports = {
    parse,
  };

File: grant_clean/6.py
# cnki parser
import re
import os
import csv
import time
import datetime
import configparser
from tools import read_tsv
from bs4 import BeautifulSoup
from multiprocessing import Pool, cpu_count

from db import sql_execute, sql_insert, sql_insert_many
from tools import write_to_csv

import random
import ssl
import urllib.request as urllib
import urllib.parse as parser

agent_file = open(os.getcwd() + '/user-agent.txt', 'r')
agent_list = agent_file.read().split('\n')
agent_file.read()

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

def request(url):
    url = parser.quote(url, safe=":?#[]!$&'()*+,;=/", encoding="utf-8")
    random_agent = agent_list[random.randint(0, len(agent_list)-1)]

    req = urllib.Request(
        url,
        headers={
            'User-Agent': random_agent
        }
    )

    f = urllib.urlopen(req, context=ctx, timeout=10)
    # print(f.getcode(), url)

    return f.read().decode()

def get_html(url, tag="", class_id="", class_id_name="") :
    html = request(url)
    soup = BeautifulSoup(html, "html.parser")
    del html
    
    try :
        if tag == "" and class_id == "" and class_id_name == "" :
            return soup
        else :
            soup = soup.find(tag, {class_id : class_id_name})
    except :
        return "NOT_FOUND"

    return soup

def cleaner(row) :
    features = {
        "id"                : row["id"],
        "grant_sites_id"    : row["grant_sites_id"],
        "project_id"        : None,
        "project_id_original" : None,
        "title"             : None,
        "title_en"          : None,
        "project_url"       : row["project_url"],
        "keyword"           : None,
        "keyword_en"        : None,
        "source"            : None,
        "researcher"        : None,
        "researcher_en"     : None,
        "researcher_cleaned" : None,
        "institution"       : None,
        "institution_en"       : None,
        "institution_cleaned" : None,
        "country"           : "FR",
        "currency"          : None,
        "award_amount"      : None,
        "award_amount_usd"  : None,
        "funded_date"       : None,
        "start_year"        : None,
        "end_year"          : None,
        "start_date"        : None,
        "end_date"          : None,
        "project_term"      : None,
        "description"       : None,
        "description_raw"   : None,
        "description_en"    : None,    
    }
    html = row["html"]
    del row

    if not html :
        # print("html not found")
        return features
    
    try :
        soup = BeautifulSoup(html, "html.parser")
    except :
        # print("Soup errors!")
        return features
    
    if features["project_url"].find("ProjetIA") != -1 :
        features["title"]    = get_title(soup)
        features["title_en"] = features["title"]

        features["project_id"]          = get_project_id(features["project_url"])
        features["project_id_original"] = features["project_id"]

        features["award_amount"], features["currency"] = get_award_amount_new(soup)
        features["researcher"], features["institution"] = get_researcher_new(soup)

        features["start_date"], features["end_date"] = get_date_new(soup)

        features["description"]  = get_description(soup)
        features["description_raw"] = features["description"]
        features["description_en"] = features["description"]

        features["start_year"] = get_end_year(features["start_date"])
        features["end_year"] = get_end_year(features["end_date"])
    else :
        features["title"]    = get_title(soup)
        features["title_en"] = features["title"]

        features["project_id"]          = get_project_id(features["project_url"])
        features["project_id_original"] = features["project_id"]
        
        features["institution"] = get_institution(soup)
        features["institution_en"] = features["institution"]

        features["award_amount"], features["currency"]  = get_award_amount(soup)
        features["start_date"], features["end_date"] = get_date(soup)

        features["description"]  = get_description(soup)
        features["description_raw"] = features["description"]
        features["description_en"] = features["description"]

        features["start_year"] = get_end_year(features["start_date"])
        features["end_year"] = get_end_year(features["end_date"])

    return features

def get_researcher_new(soup) :
    try :
        soup = soup.find("section", {"class" : "block"}).find("p")
    except :
        return None, None

    researcher, institution = None, None
    soup = BeautifulSoup(str(soup).replace("<br/>", ";"), "html.parser")
    
    for row in soup.get_text().split(";") :
        if row.strip() == "" :
            continue
        cand = row.split(":")
        if len(cand) == 2 :
            if cand[0].strip() == "RST" :
                researcher = cand[1].strip()
            elif cand[0].strip() == "Etablissement Coordinateur" :
                institution = cand[1].strip()
    
    return researcher, institution

def get_title(soup) :
    try :
        return soup.find("section", {"class" : "content-style"}).find("h1").get_text().strip()
    except :
        return None
    return None

def get_institution(soup) :
    try :
        institution_cand = soup.find_all("section", {"class" : "block block-info"})
    except :
        return None
    # print(institution_cand)
    try :
        for row in institution_cand :
            if row.find("h2").get_text().strip() == "Partner" :
                institution = row.find("div", {"class" : "block__content"}).find("p").get_text()
                institution = ";".join([ins.strip() for ins in institution.split("\n") if ins.strip() != ""])
                return institution
    except : 
        return None
    return None

def get_description(soup) :
    try :
        description = soup.find("section", {"class" : "content-style"})
    except :
        return None
    [s.extract() for s in description.find_all("section")]
    [s.extract() for s in description.find_all("h1")]
    [s.extract() for s in description.find_all("div", {"class" : "categories"})]

    return description.get_text().strip()

def get_award_amount(soup) :
    try :
        cands = soup.find_all("section", {"class" : "block block-info"})
    except :
        return None, None

    try :
        for row in cands :
            if row.find("h2").get_text().strip() == "Partner" :
                strongs = row.find_all("strong")
                for strong in strongs :
                    if strong.get_text().find("ANR") != -1 :
                        currency = "EUR" if strong.find("euros") != -1 else None
                        award_amount = re.sub(r"[^\d\.]", "", strong.get_text())
                        try :
                            award_amount = float(award_amount)
                        except :
                            return None, None
                        return award_amount, currency
    except : 
        return None, None
    return None, None

def get_award_amount_new(soup) :
    try :
        soup = soup.find("section", {"class" : "block"})
    except :
        return None, None

    try :
        strongs = soup.find_all("strong")
        for strong in strongs :
            if strong.get_text().find("ANR") != -1 :
                currency = "EUR" if strong.find("euros") != -1 else None
                award_amount = re.sub(r"[^\d\.]", "", strong.get_text())
                try :
                    award_amount = float(award_amount)
                except :
                    return None, None
                return award_amount, currency
    except : 
        return None, None
    return None, None

def get_date_new(soup) :
    try :
        soup = soup.find("section", {"class" : "block"})
    except :
        return None, None
    
    try :
        strongs = soup.find_all("strong")
        for strong in strongs :
            if strong.get_text().find("Investissement") != -1 :
                text = strong.get_text().strip()
                text = text.split("période de")[-1]
                cal = {
                    "janvier" : "January",
                    "février" : "February",
                    "mars" : "March",
                    "avril" : "April",
                    "mai" : "May",
                    "juin" : "June",
                    "juillet" : "July",
                    "aout" : "August",
                    "septembre" : "September",
                    "octobre" : "October",
                    "novembre" : "November",
                    "décembre" : "December",
                }
                tmp = []
                for row in text.split() :
                    if row.lower().strip() in cal :
                        tmp.append(cal[row.lower().strip()])
                    else :
                        tmp.append(row)
                text = " ".join(tmp)
                text = [t.strip() for t in text.split("à")]
                if len(text) > 1 :
                    if text[0] == "" :
                        return None, None
                    text[0] = datetime.datetime.strptime(text[0], "%B %Y").date()
                    text[1] = datetime.datetime.strptime(text[1], "%B %Y").date()
                    return str(text[0]), str(text[1])
    except : 
        return None, None
    return None, None

def get_date(soup) :
    try :
        cands = soup.find_all("section", {"class" : "block block-info"})
    except :
        return None, None
    
    try :
        for row in cands :
            if row.find("h2").get_text().strip() == "Partner" :
                strongs = row.find_all("strong")
                for strong in strongs :
                    if strong.get_text().find("project") != -1 :
                        text = strong.get_text().strip()
                        text = text.split(":")[-1]
                        text = [t.strip() for t in text.split("-")]
                        if len(text) > 1 :
                            if text[0] == "" :
                                return None, None
                            text[0] = datetime.datetime.strptime(text[0], "%B %Y").date()
                            tmp = text[0]
                            pos = text[1].find(" Months")
                            
                            if pos != -1 :
                                year_save = int(int(text[1][: pos]) / 12)
                                month_save = int(int(text[1][: pos]) % 12)
                                
                                year_save = tmp.year + year_save + int((tmp.month + month_save - 1) / 12)
                                month_save = (month_save + tmp.month) % 12
                                if month_save == 0 :
                                    month_save = 12

                                # text[1] = tmp.replace(year=tmp.year + int(int(text[1][:2])/12), month=tmp.month + int(int(text[1][:2]) % 12))
                                text[1] = tmp.replace(year=year_save, month=month_save)
                                
                                return str(text[0]), str(text[1])
    except : 
        return None, None
    return None, None

def get_end_year(end_date) :
    if end_date is None :
        return None
    if len(end_date) > 4 :
        return end_date[ : 4]
    return None

def get_project_id(project_url) :
    if project_url is None :
        return None
    if project_url.find("-") != -1 :
        return project_url[project_url.find("-") + 1 :].strip()
    return None

def multi_process_db() :
    start_time = time.time()

    sql = """SELECT * FROM tmp_anr_html WHERE html is not null"""
    query_result = sql_execute(sql, db_config)

    sub_grant_datas = []

    for row in query_result :
        features = {
            "id"              : row[0],
            "grant_sites_id"  : row[1],
            "project_url"     : row[2],
            "html"            : row[3],
        }
        sub_grant_datas.append(features)

    p = Pool(processes=cpu_count())
    grant_datas = p.map(cleaner, sub_grant_datas)
    p.close()
    p.join()

    del sub_grant_datas
    sql = """INSERT INTO tmp_anr_info(
        id,
        grant_sites_id,
        project_id,
        project_id_original,
        title,
        title_en,
        project_url,
        keyword,
        keyword_en,
        source,
        researcher,
        researcher_en,
        researcher_cleaned,
        institution,
        institution_en,
        institution_cleaned,
        country,
        currency,
        award_amount,
        award_amount_usd,
        funded_date,
        start_year,
        end_year,
        start_date,
        end_date,
        project_term,
        description,
        description_raw,
        description_en,
        created_date,
        updated_date,
        crawled_date)
        VALUES(%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s, now(),now(),now() )"""
    
    input_datas = []
    
    for row in grant_datas :
        input_datas.append((
            int(row["id"]),
            int(row["grant_sites_id"]),
            row["project_id"],
            row["project_id_original"],
            row["title"],
            row["title_en"],
            row["project_url"],
            row["keyword"],
            row["keyword_en"],
            row["source"],
            row["researcher"],
            row["researcher_en"],
            row["researcher_cleaned"],
            row["institution"],
            row["institution_en"],
            row["institution_cleaned"],
            row["country"],
            row["currency"],
            row["award_amount"],
            row["award_amount_usd"],
            row["funded_date"],
            row["start_year"],
            row["end_year"],
            row["start_date"],
            row["end_date"],
            row["project_term"],
            row["description"],
            row["description_raw"],
            row["description_en"]
        ))

    sql_insert_many(sql, db_config, input_datas)

if __name__ == "__main__":
    start_time = time.time()

    config = configparser.ConfigParser()
    config.read('.env')

    db_config = (config["DEFAULT"]["_HOSTNAME"], config["DEFAULT"]["_DATABASE"], config["DEFAULT"]["_USERNAME"], \
    config["DEFAULT"]["_PASSWORD"], config["DEFAULT"]["_PORT"])

    multi_process_db()

    # row = {
    #     "id" : 1,
    #     "grant_sites_id" : 6,
    #     "project_url" : "https://anr.fr/Project-ANR-13-BIME-0005"
    # }
    # with open("./tmp/test.html") as f :
    #     row["html"] = f.read()

    # tmp = cleaner(row)

    # for key, val in tmp.items() :
    #     print(key, val)

    print("Total time is : ", str(time.time() - start_time))